src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs

```diff
@@ -1,161 +1,132 @@
+        match = Regex.Match($"{testedPath}/", regexPattern);
+        return match.Groups.Cast<Group>().Skip(1).ToList();
+    }
 
-        private static int GetNextCounterPosition(string urlTemplate, int counterForTemplate, char delimiter)
-        {
-            var closingPlaceHolderPositionOnTemplate = urlTemplate.IndexOf(delimiter, counterForTemplate);
-            return closingPlaceHolderPositionOnTemplate + 1;
-        }
+    /// <summary>
+    /// Checks if the path template contains a catch-all query parameter.
+    /// It means that the path template ends with a question mark and a placeholder.
+    /// And no other placeholders are present in the path template.
+    /// </summary>
+    /// <param name="template">The path template.</param>
+    /// <returns>True if it matches and the found placeholder.</returns>
+    private static (bool IsMatch, string Placeholder) IsCatchAllQuery(string template)
+    {
+        Regex catchAllQueryRegex = new($@"^[^{{}}]*\?\{PlaceHolderLeft}(.*?)\{PlaceHolderRight}$", RegexOptions.None, TimeSpan.FromMilliseconds(300));
+        Match catchAllMatch = catchAllQueryRegex.Match(template);
 
-        private static bool CharactersDontMatch(char characterOne, char characterTwo)
-        {
-            return char.ToLower(characterOne) != char.ToLower(characterTwo);
-        }
+        return (catchAllMatch.Success, catchAllMatch.Success ? catchAllMatch.Groups[1].Value : string.Empty);
+    }
 
-        private static bool ContinueScanningUrl(int counterForUrl, int urlLength)
+    /// <summary>
+    /// Check if the path template contains a catch-all path parameter.
+    /// It means that the path template ends with a placeholder and no other placeholders are present in the path template,
+    /// without a question mark (query parameters).
+    /// </summary>
+    /// <param name="template">The path template.</param>
+    /// <returns>True if it matches.</returns>
+    private static bool IsCatchAllPath(string template)
+    {
+        Regex catchAllPathRegex = new($@"^[^{{}}]*\{PlaceHolderLeft}(.*?)\{PlaceHolderRight}/?$", RegexOptions.None, TimeSpan.FromMilliseconds(300));
+        return catchAllPathRegex.IsMatch(template) && !template.Contains('?');
+    }
+    
+    /// <summary>
+    /// Checks if the query should be skipped.
+    /// It should be skipped if it is not present in the path template.
+    /// </summary>
+    /// <param name="query">The query string.</param>
+    /// <param name="template">The path template.</param>
+    /// <returns>True if query should be skipped.</returns>
+    private static bool ShouldSkipQuery(string query, string template) => !string.IsNullOrEmpty(query) && !template.Contains('?');
+
+    /// <summary>
+    /// Escapes all characters except braces, eg { and }.
+    /// </summary>
+    /// <param name="input">The input string.</param>
+    /// <returns>The formatted string.</returns>
+    private static string EscapeExceptBraces(string input)
```

Comment:
I really wonder about this method. Why do we have to escape like this?

Reply:
For performance sake...

Reply:
Not an issue